test/lib/jdk/test/lib/net/HttpHeaderParser.java

```diff
@@ -80,4 +109,278 @@ public HttpHeaderParser(InputStream is) throws IOException {
+     * ( when true is returned ), which may not be fully consumed.
+     *
+     * @param input the ( partial ) header data
+     * @return true iff the end of the headers block has been reached
+     */
+    public boolean parse(InputStream input) throws IOException {
+        requireNonNull(input, "null input");
+        while (canContinueParsing(input)) {
+            switch (state) {
+                case INITIAL                                    ->  state = HttpHeaderParser.State.STATUS_LINE;
+                case STATUS_LINE                                ->  readResumeStatusLine(input);
+                case STATUS_LINE_FOUND_CR, STATUS_LINE_FOUND_LF ->  readStatusLineFeed(input);
+                case STATUS_LINE_END                            ->  maybeStartHeaders(input);
+                case STATUS_LINE_END_CR, STATUS_LINE_END_LF     ->  maybeEndHeaders(input);
+                case HEADER                                     ->  readResumeHeader(input);
+                case HEADER_FOUND_CR, HEADER_FOUND_LF           ->  resumeOrLF(input);
+                case HEADER_FOUND_CR_LF                         ->  resumeOrSecondCR(input);
+                case HEADER_FOUND_CR_LF_CR                      ->  resumeOrEndHeaders(input);
+                default -> throw new InternalError("Unexpected state: " + state);
+            }
+        }
+        return state == HttpHeaderParser.State.FINISHED;
+    }
+
+    private boolean canContinueParsing(InputStream buffer) throws IOException {
+        // some states don't require any input to transition
+        // to the next state.
+        return switch (state) {
+            case FINISHED -> false;
+            case STATUS_LINE_FOUND_LF, STATUS_LINE_END_LF, HEADER_FOUND_LF -> true;
+            default -> buffer.available() >= 0;
+        };
+    }
+
+    /**
+     * Returns a character (char) corresponding to the next byte in the
+     * input, interpreted as an ISO-8859-1 encoded character.
+     * <p>
+     * The ISO-8859-1 encoding is a 8-bit character coding that
+     * corresponds to the first 256 Unicode characters - from U+0000 to
+     * U+00FF. UTF-16 is backward compatible with ISO-8859-1 - which
+     * means each byte in the input should be interpreted as an unsigned
+     * value from [0, 255] representing the character code.
+     *
+     * @param input a {@code ByteBuffer} containing a partial input
+     * @return the next byte in the input, interpreted as an ISO-8859-1
+     * encoded char
+     * @throws BufferUnderflowException
+     *          if the input buffer's current position is not smaller
+     *          than its limit
+     */
+    private char get(InputStream input) throws IOException {
+        return (char)(input.read() & 0xFF);
+    }
+
+    private void readResumeStatusLine(InputStream input) throws IOException {
+        char c = 0;
+        while (input.available() > 0 && (c = get(input)) != CR) {
```

Comment:
It's not a too good idea to depend on input.available() - the spec leaves too much room for odd behaviors.
https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/io/InputStream.html#available()
An implementation may decide to return 0 always - as it is the amount of bytes that is guaranteed to be readable without blocking (and an implementation may not know how many bytes can be read without blocking).